scripts/init.sh

```diff
@@ -3,92 +3,111 @@
 
 set -e
 
-case "$(uname -s):$(uname -m):$(uname -v):$(uname -a)" in
-Linux:x86_64*)     
-    TARGET="x86_64-unknown-linux-musl"
-    ;;
-Linux:aarch64:*:*Android*)
-    TARGET="aarch64-linux-android"
-    ;;
-Linux:arm64*|Linux:aarch64*)
-    TARGET="aarch64-unknown-linux-musl"
-    ;;
-Linux:arm*)     
-    TARGET="armv7-unknown-linux-musleabihf"
-    ;;
-Linux:i686:iSH*)
-    TARGET="ish"
-    ;;
-Linux:i686*)     
-    TARGET="i686-unknown-linux-musl"
-    ;;
-Linux:i586*)     
-    TARGET="i586-unknown-linux-musl"
-    ;;
-Linux:mips*)     
-    TARGET="mips-unknown-linux-musl"
-    ;;
-FreeBSD:x86_64*)
-    TARGET="x86_64-unknown-freebsd"
-    ;;
-FreeBSD:amd64*)
-    TARGET="x86_64-unknown-freebsd"
-    ;;
-FreeBSD:i686*)
-    TARGET="i686-unknown-freebsd"
-    ;;
-Darwin:x86_64*)    
-    TARGET="x86_64-apple-darwin"
-    ;;
-WindowsNT:x86_64*)    
-    TARGET="x86_64-pc-windows-msvc.exe"
-    ;;
-WindowsNT:i686*)    
-    TARGET="i686-pc-windows-msvc.exe"
-    ;;
-*)          
-    echo "Unsupported system ($(uname -a))"
-    exit 1
-    ;;
-esac
+main() {
+    case "$(uname -s):$(uname -m):$(uname -v):$(uname -a)" in
+    Linux:x86_64*)     
+        TARGET="x86_64-unknown-linux-musl"
+        ;;
+    Linux:aarch64:*:*Android*)
+        TARGET="aarch64-linux-android"
+        ;;
+    Linux:arm64*|Linux:aarch64*)
+        TARGET="aarch64-unknown-linux-musl"
+        ;;
+    Linux:arm*)     
+        TARGET="armv7-unknown-linux-musleabihf"
+        ;;
+    Linux:i686:iSH*)
+        TARGET="ish"
+        ;;
+    Linux:i686*)     
+        TARGET="i686-unknown-linux-musl"
+        ;;
+    Linux:i586*)     
+        TARGET="i586-unknown-linux-musl"
+        ;;
+    Linux:mips*)     
+        TARGET="mips-unknown-linux-musl"
+        ;;
+    FreeBSD:x86_64*)
+        TARGET="x86_64-unknown-freebsd"
+        ;;
+    FreeBSD:amd64*)
+        TARGET="x86_64-unknown-freebsd"
+        ;;
+    FreeBSD:i686*)
+        TARGET="i686-unknown-freebsd"
+        ;;
+    Darwin:x86_64*)    
+        TARGET="x86_64-apple-darwin"
+        ;;
+    WindowsNT:x86_64*)    
+        TARGET="x86_64-pc-windows-msvc.exe"
+        ;;
+    WindowsNT:i686*)    
+        TARGET="i686-pc-windows-msvc.exe"
+        ;;
+    *)          
+        echo "Unsupported system ($(uname -a))"
+        exit 1
+        ;;
+    esac
 
-if [ -z "$TMPDIR" ]
-then
-    TMPDIR="/tmp"
-fi
-
-TEMP_PATH="$TMPDIR/tunshell"
-CLIENT_PATH="$TEMP_PATH/client"
-HEADERS_PATH="$TEMP_PATH/headers"
+    if [ ! -z "$XDG_CACHE_HOME" ]
+    then
+        TEMP_PATH="$XDG_CACHE_HOME"
```

Comment:
While `$XDG_CACHE_HOME` is unset you are supposed to default to ` $HOME/.cache`, I think it is quite rate for `$XDG_CACHE_HOME` to be explicitly set. This provides better assurance that the directory is created in a location that the user controls. However I am not sure how well that works for daemons without a home or docker containers.

Reply:
Don't want to make too many assumptions about the host executing the script. I've modified the script to use the following paths:

1. `XDG_CACHE_HOME` if it is set
2. `TMPDIR` if it is set or writable
3. `/tmp` if it writable
4. `mktemp` if it available

Reply:
I would:

- Change 1 to be "if is writable" instead of "if is set" and follow the correct fallback logic. 
- Don't use 2 or 3 with an existing directory. Always create a new directory with a random name. I would also do these after 4
- 4 This is good. I would do this second if `XDG_CACHE_HOME` is not writable.

Reply:
I've cleaned up the checks to use `-w` as mentioned but I've kept the ordering as-is since using `mktemp` would prevent us caching the client for future use.
To mitigate the risk i've updated the caching to only use the cached client if it's `md5sum` is equal to that returned from S3 regardless of the chosen temp directory.

Reply:
I think that reusing a client that happens to exist in a world-writable directory like `/tmp` is fundamentally risky and think that the downsides do not justify the benefits. My recommendations would be:

1. It is safe to use a binary if it exists in `$HOME` or `$XDG_CACHE_HOME`.
2. It is safe to use a binary that has been copied from anywhere and verified against a strong hash or signature (md5 does not qualify in my mind).
    - This includes HTTPS verified from a trusted host.
3. It is not safe to use a binary that lives inside of a world writable directory (like `/tmp`) that can not be verified to be created by the user.

Reply:
Agree that md5 is not the right choice. I've updated this to use sha256.
Additionally also to use `$HOME/.cache` ahead of tmp dirs if it's accessible.